### ISSUE TITLE
CA1859 erroneously recommends nullable return type

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#6413]
### Analyzer

**Diagnostic ID**: CA1859

### Analyzer source

https://github.com/dotnet/roslyn-analyzers/commit/6ae2683c5033dfde0b73d4e73b372f00ca6ceacd

### Describe the bug

The analyzer recommends changing a non-nullable return type to a nullable one.

### Steps To Reproduce

```C#
[Fact]
public static async Task Repro()
{
    const string Source = @"
        #nullable enable

        using System;
        using System.IO;

        namespace Example
        {
            internal class C
            {
                private MemoryStream? _stream;

                private Stream GetStream()
                {
                    return _stream ?? Create();
                }

                private MemoryStream Create() => new MemoryStream();
            }
        }
        ";

    await TestCSAsync(Source);
}
```

### Expected behavior

```
info CA1859: Change return type of method 'GetStream' from 'System.IO.Stream' to 'System.IO.MemoryStream' for improved performance
```

### Actual behavior

```
info CA1859: Change return type of method 'GetStream' from 'System.IO.Stream' to 'System.IO.MemoryStream?' for improved performance
```

cc: @geeknoid, @buyaa-n

**Comment by Youssef1313** [comment]
Closing as fixed per https://github.com/dotnet/roslyn-analyzers/pull/6418#issuecomment-1385520293
